src/rate.rs

```diff
@@ -0,0 +1,218 @@
+use anyhow::{bail, Context};
+use num::{BigUint, Integer, ToPrimitive};
+use std::convert::TryFrom;
+use std::iter::FromIterator;
+use std::str::FromStr;
+
+/// Represent a rate. Note this is designed to support Bitcoin/Dai buy and sell rates (Bitcoin being in the range of 10k-100kDai)
+/// A rate has a maximum precision of 9 digits after the decimal
+// rate = self.0 * 10e-9
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Rate(u64);
+
+impl Rate {
+    pub const PRECISION: u16 = 9;
+    const MAX_RATE: u64 = (u64::MAX / 10) ^ 9;
+
+    /// integer = rate * 10ePRECISION
+    pub fn new(integer: u64) -> anyhow::Result<Self> {
+        if integer > Self::MAX_RATE {
+            bail!("Rate is larger than supported");
+        }
+        Ok(Rate(integer))
+    }
+
+    /// integer = rate * 10ePRECISION
+    pub fn integer(&self) -> BigUint {
+        BigUint::from(self.0)
+    }
+}
+
+impl TryFrom<f64> for Rate {
+    type Error = anyhow::Error;
+
+    fn try_from(rate: f64) -> Result<Self, Self::Error> {
+        if rate.is_sign_negative() {
+            anyhow::bail!("Rate must be positive");
+        }
+
+        if !rate.is_finite() {
+            anyhow::bail!("Rate must be finite")
+        }
+
+        let mut rate = rate.to_string();
+        let decimal_index = rate.find('.');
+        let mantissa = match decimal_index {
+            None => String::new(),
+            Some(decimal_index) => {
+                let mantissa = rate.split_off(decimal_index + 1);
+                if mantissa.len() > Self::PRECISION as usize {
+                    anyhow::bail!(format!(
+                        "Precision of the rate is too high (max is {}).",
+                        Self::PRECISION
+                    ))
+                } else {
+                    rate.truncate(rate.len() - 1); // Removes the trailing decimal point
+                    mantissa
+                }
+            }
+        };
+
+        let mantissa_length = mantissa.len();
+        let integer = rate;
+
+        let zeros = vec!['0'].repeat(Self::PRECISION as usize - mantissa_length);
+        let zeros = String::from_iter(zeros.into_iter());
+        let integer = u64::from_str(&format!("{}{}{}", integer, mantissa, zeros))
+            .context("Rate is unexpectedly large")?;
+        Rate::new(integer)
+    }
+}
+
+/// Spread: percentage to be added on top of a rate or amount with
+/// a maximum precision of 2 decimals
+#[derive(Clone, Copy, Debug)]
+pub struct Spread(u16);
+
+impl Spread {
+    /// Input is the spread in percent, with 2 digits after the decimal point:
```

Comment:
If I wouldn't know what you mean, then I would have trouble understanding this explanation :sweat_smile:. (There are no decimal numbers here.)

The technical term for this unit seems to be `permyriad`: https://en.wikipedia.org/wiki/Basis_point#Permyriad for "1 in 10_000".

I've never heard of this, so it might be worthwhile to think about using something more common like [`per cent mille`](https://en.wikipedia.org/wiki/Per_cent_mille) or [`parts per million`](https://en.wikipedia.org/wiki/Parts-per_notation).

Maybe "parts per ten-thousand" would also be sufficient as an explanation.

Reply:
If we don't want to bother the end-user with this language, we can always introduce a translation layer in the config file that talks about "per cent".

I see this function as part of the application core. If we accept permyriad in the argument, then I would also use that language in the documentation and the error messages for consistency.

Reply:
Thanks. Yes I'd expect the config to take a float percent.